### ISSUE TITLE
criterion:backward(output, target) why output and target not in the same dimension

### ISSUE THREAD

**Issue by hjyCoder** [Element-Research/rnn#99]
When I see the code of Example, I don't understand that why output and target not in the same dimension on the loss function ?

**Comment by nicholas-leonard** [comment]
@hjyCoder Could you be more specific. Which example, which line? :)

**Comment by hjyCoder** [comment]
-- the example is below. the line of  "  local gradOutput = criterion:backward(output, target) "

require 'rnn'

batchSize = 8
rho = 5
hiddenSize = 10
nIndex = 10000
-- RNN
r = nn.Recurrent(
   hiddenSize, nn.LookupTable(nIndex, hiddenSize), 
   nn.Linear(hiddenSize, hiddenSize), nn.Sigmoid(), 
   rho
)

rnn = nn.Sequential()
rnn:add(r)
rnn:add(nn.Linear(hiddenSize, nIndex))
rnn:add(nn.LogSoftMax())

criterion = nn.ClassNLLCriterion()

-- dummy dataset (task is to predict next item, given previous)
sequence = torch.randperm(nIndex)

offsets = {}
for i=1,batchSize do
   table.insert(offsets, math.ceil(math.random()*batchSize))
end
offsets = torch.LongTensor(offsets)

lr = 0.1
updateInterval = 4
i = 1
while true do
   -- a batch of inputs
   local input = sequence:index(1, offsets)
   local output = rnn:forward(input)
   -- incement indices
   offsets:add(1)
   for j=1,batchSize do
      if offsets[j] > nIndex then
         offsets[j] = 1
      end
   end
   local target = sequence:index(1, offsets)
   local err = criterion:forward(output, target)
   print(err)
   local gradOutput = criterion:backward(output, target)
   -- the Recurrent layer is memorizing its gradOutputs (up to memSize)
   rnn:backward(input, gradOutput)

   i = i + 1
   -- note that updateInterval < rho
   if i % updateInterval == 0 then
      -- backpropagates through time (BPTT) :
      -- 1. backward through feedback and input layers,
      rnn:backwardThroughTime()
      -- 2. updates parameters
      rnn:updateParameters(lr)
      rnn:zeroGradParameters()
      -- 3. reset the internal time-step counter
      rnn:forget()
   end
end

**Comment by nicholas-leonard** [comment]
@hjyCoder What do you mean by "not in the same dimension of loss function" ?

**Comment by hjyCoder** [comment]
In the example, output is the [torch.DoubleTensor of size 8x10000], but the target is the [torch.DoubleTensor of size 8], which configure parameters make it know change to each other, then computer the function loss.

**Comment by nicholas-leonard** [comment]
@hjyCoder Oh I see. Yeah so the [ClassNLLCriterion](https://github.com/torch/nn/blob/master/doc/criterion.md#nn.ClassNLLCriterion) expects the log-likelihood over all classes as input (in this case, a vector of 10000 classes), and the class index as target ( a scalar index between 1 and 10000).

**Comment by hughperkins** [comment]
(Note: @hjyCoder if you used MSECriterion , then output and target would be in the same dimension, but then you'd have to 'unpack' the targets value into 1-of-k vectors)

**Comment by hjyCoder** [comment]
@nicholas-leonard , I want to train the RNN to do the function that predict the next number, example input = { 1,2,3,4}, target = {2,3,4,5}. I change the core code like below:
"
-- a batch of inputs
    local input = sequence:index(1, offsets)
    local target = sequence:index(1, offsets)
    target = target:add(1)
    local output = rnn:forward(input)
    -- incement indices
    offsets:add(1)
    for j=1,batchSize do
        if offsets[j] > nIndex then
            offsets[j] = 1
        end
    end
   -- local target = sequence:index(1, offsets)
    local err = criterion:forward(output, target)
    print(err)
    local gradOutput = criterion:backward(output, target)
    -- the Recurrent layer is memorizing its gradOutputs (up to memSize)
    rnn:backward(input, gradOutput)
"
why I got the error like that:
/Users/huangjiyang/torch/install/bin/luajit: ...ang/torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:47: Assertion `cur_target >= 0 && cur_target < n_classes' failed.  at /tmp/luarocks_nn-scm-1-1588/nn/generic/ClassNLLCriterion.c:52
stack traceback:
	[C]: in function 'ClassNLLCriterion_updateOutput'
	...ang/torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:47: in function 'forward'
	test6.lua:50: in main chunk
	[C]: in function 'dofile'
	...yang/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x01072dc4c0

**Comment by nicholas-leonard** [comment]
The error occurs when the target class is greater than the size of the output (likelihood of classes). You are incrementing the targets. You should instead increment the offsets. Try this : 
```lua
    -- a batch of inputs
    local input = sequence:index(1, offsets)
    local output = rnn:forward(input)
    -- incement indices
    offsets:add(1)
    for j=1,batchSize do
        if offsets[j] > nIndex then
            offsets[j] = 1
        end
    end
    local target = sequence:index(1, offsets)
    local err = criterion:forward(output, target)
    print(err)
    local gradOutput = criterion:backward(output, target)
    -- the Recurrent layer is memorizing its gradOutputs (up to memSize)
    rnn:backward(input, gradOutput)
```

**Comment by hjyCoder** [comment]
@nicholas-leonard , thank you for your help.
